In [156]:
import pandas as pd

In [157]:
data1=pd.read_csv(r'C:\Users\kalali aravindgoud\Desktop\Dataset\True.csv')

In [158]:
data1

title  \
0      As U.S. budget fight looms, Republicans flip t...   
1      U.S. military to accept transgender recruits o...   
2      Senior U.S. Republican senator: 'Let Mr. Muell...   
3      FBI Russia probe helped by Australian diplomat...   
4      Trump wants Postal Service to charge 'much mor...   
...                                                  ...   
21412  'Fully committed' NATO backs new U.S. approach...   
21413  LexisNexis withdrew two products from Chinese ...   
21414  Minsk cultural hub becomes haven from authorities   
21415  Vatican upbeat on possibility of Pope Francis ...   
21416  Indonesia to buy $1.14 billion worth of Russia...   

                                                    text       subject  \
0      WASHINGTON (Reuters) - The head of a conservat...  politicsNews   
1      WASHINGTON (Reuters) - Transgender people will...  politicsNews   
2      WASHINGTON (Reuters) - The special counsel inv...  politicsNews   
3      WASHINGTON (Reuters) - Trump campaign adviser ...  politicsNews   
4      SEATTLE/WASHINGTON (Reuters) - President Donal...  politicsNews   
...                                                  ...           ...   
21412  BRUSSELS (Reuters) - NATO allies on Tuesday we...     worldnews   
21413  LONDON (Reuters) - LexisNexis, a provider of l...     worldnews   
21414  MINSK (Reuters) - In the shadow of disused Sov...     worldnews   
21415  MOSCOW (Reuters) - Vatican Secretary of State ...     worldnews   
21416  JAKARTA (Reuters) - Indonesia will buy 11 Sukh...     worldnews   

                     date  
0      December 31, 2017   
1      December 29, 2017   
2      December 31, 2017   
3      December 30, 2017   
4      December 29, 2017   
...                   ...  
21412    August 22, 2017   
21413    August 22, 2017   
21414    August 22, 2017   
21415    August 22, 2017   
21416    August 22, 2017   

[21417 rows x 4 columns]

In [159]:
data1['statement']='True'

In [160]:
data1

title  \
0      As U.S. budget fight looms, Republicans flip t...   
1      U.S. military to accept transgender recruits o...   
2      Senior U.S. Republican senator: 'Let Mr. Muell...   
3      FBI Russia probe helped by Australian diplomat...   
4      Trump wants Postal Service to charge 'much mor...   
...                                                  ...   
21412  'Fully committed' NATO backs new U.S. approach...   
21413  LexisNexis withdrew two products from Chinese ...   
21414  Minsk cultural hub becomes haven from authorities   
21415  Vatican upbeat on possibility of Pope Francis ...   
21416  Indonesia to buy $1.14 billion worth of Russia...   

                                                    text       subject  \
0      WASHINGTON (Reuters) - The head of a conservat...  politicsNews   
1      WASHINGTON (Reuters) - Transgender people will...  politicsNews   
2      WASHINGTON (Reuters) - The special counsel inv...  politicsNews   
3      WASHINGTON (Reuters) - Trump campaign adviser ...  politicsNews   
4      SEATTLE/WASHINGTON (Reuters) - President Donal...  politicsNews   
...                                                  ...           ...   
21412  BRUSSELS (Reuters) - NATO allies on Tuesday we...     worldnews   
21413  LONDON (Reuters) - LexisNexis, a provider of l...     worldnews   
21414  MINSK (Reuters) - In the shadow of disused Sov...     worldnews   
21415  MOSCOW (Reuters) - Vatican Secretary of State ...     worldnews   
21416  JAKARTA (Reuters) - Indonesia will buy 11 Sukh...     worldnews   

                     date statement  
0      December 31, 2017       True  
1      December 29, 2017       True  
2      December 31, 2017       True  
3      December 30, 2017       True  
4      December 29, 2017       True  
...                   ...       ...  
21412    August 22, 2017       True  
21413    August 22, 2017       True  
21414    August 22, 2017       True  
21415    August 22, 2017       True  
21416    August 22, 2017       True  

[21417 rows x 5 columns]

###### Above dataset `data1`contains all the True reviews of having 21417 rows

In [161]:
data2=pd.read_csv(r'C:\Users\kalali aravindgoud\Desktop\Dataset\Fake.csv')

In [162]:
data2

title  \
0       Donald Trump Sends Out Embarrassing New Year’...   
1       Drunk Bragging Trump Staffer Started Russian ...   
2       Sheriff David Clarke Becomes An Internet Joke...   
3       Trump Is So Obsessed He Even Has Obama’s Name...   
4       Pope Francis Just Called Out Donald Trump Dur...   
...                                                  ...   
23476  McPain: John McCain Furious That Iran Treated ...   
23477  JUSTICE? Yahoo Settles E-mail Privacy Class-ac...   
23478  Sunnistan: US and Allied ‘Safe Zone’ Plan to T...   
23479  How to Blow $700 Million: Al Jazeera America F...   
23480  10 U.S. Navy Sailors Held by Iranian Military ...   

                                                    text      subject  \
0      Donald Trump just couldn t wish all Americans ...         News   
1      House Intelligence Committee Chairman Devin Nu...         News   
2      On Friday, it was revealed that former Milwauk...         News   
3      On Christmas day, Donald Trump announced that ...         News   
4      Pope Francis used his annual Christmas Day mes...         News   
...                                                  ...          ...   
23476  21st Century Wire says As 21WIRE reported earl...  Middle-east   
23477  21st Century Wire says It s a familiar theme. ...  Middle-east   
23478  Patrick Henningsen  21st Century WireRemember ...  Middle-east   
23479  21st Century Wire says Al Jazeera America will...  Middle-east   
23480  21st Century Wire says As 21WIRE predicted in ...  Middle-east   

                    date  
0      December 31, 2017  
1      December 31, 2017  
2      December 30, 2017  
3      December 29, 2017  
4      December 25, 2017  
...                  ...  
23476   January 16, 2016  
23477   January 16, 2016  
23478   January 15, 2016  
23479   January 14, 2016  
23480   January 12, 2016  

[23481 rows x 4 columns]

In [163]:
data2['statement']='fake'

In [164]:
data2

title  \
0       Donald Trump Sends Out Embarrassing New Year’...   
1       Drunk Bragging Trump Staffer Started Russian ...   
2       Sheriff David Clarke Becomes An Internet Joke...   
3       Trump Is So Obsessed He Even Has Obama’s Name...   
4       Pope Francis Just Called Out Donald Trump Dur...   
...                                                  ...   
23476  McPain: John McCain Furious That Iran Treated ...   
23477  JUSTICE? Yahoo Settles E-mail Privacy Class-ac...   
23478  Sunnistan: US and Allied ‘Safe Zone’ Plan to T...   
23479  How to Blow $700 Million: Al Jazeera America F...   
23480  10 U.S. Navy Sailors Held by Iranian Military ...   

                                                    text      subject  \
0      Donald Trump just couldn t wish all Americans ...         News   
1      House Intelligence Committee Chairman Devin Nu...         News   
2      On Friday, it was revealed that former Milwauk...         News   
3      On Christmas day, Donald Trump announced that ...         News   
4      Pope Francis used his annual Christmas Day mes...         News   
...                                                  ...          ...   
23476  21st Century Wire says As 21WIRE reported earl...  Middle-east   
23477  21st Century Wire says It s a familiar theme. ...  Middle-east   
23478  Patrick Henningsen  21st Century WireRemember ...  Middle-east   
23479  21st Century Wire says Al Jazeera America will...  Middle-east   
23480  21st Century Wire says As 21WIRE predicted in ...  Middle-east   

                    date statement  
0      December 31, 2017      fake  
1      December 31, 2017      fake  
2      December 30, 2017      fake  
3      December 29, 2017      fake  
4      December 25, 2017      fake  
...                  ...       ...  
23476   January 16, 2016      fake  
23477   January 16, 2016      fake  
23478   January 15, 2016      fake  
23479   January 14, 2016      fake  
23480   January 12, 2016      fake  

[23481 rows x 5 columns]

###### In another  dataset `data2` contains all the  Fake reviews of having 23481 rows

In [165]:
import numpy as np

In [166]:
data1.isnull().sum()

title        0
text         0
subject      0
date         0
statement    0
dtype: int64

In [167]:
data2.isnull().sum()

title        0
text         0
subject      0
date         0
statement    0
dtype: int64

In [168]:
data1['subject'].value_counts()

politicsNews    11272
worldnews       10145
Name: subject, dtype: int64

In [169]:
data2['subject'].value_counts()

News               9050
politics           6841
left-news          4459
Government News    1570
US_News             783
Middle-east         778
Name: subject, dtype: int64

In [170]:
data=pd.concat([data1,data2],axis=0) 

###### Concating two dataframes with having identical column names

In [171]:
data

title  \
0      As U.S. budget fight looms, Republicans flip t...   
1      U.S. military to accept transgender recruits o...   
2      Senior U.S. Republican senator: 'Let Mr. Muell...   
3      FBI Russia probe helped by Australian diplomat...   
4      Trump wants Postal Service to charge 'much mor...   
...                                                  ...   
23476  McPain: John McCain Furious That Iran Treated ...   
23477  JUSTICE? Yahoo Settles E-mail Privacy Class-ac...   
23478  Sunnistan: US and Allied ‘Safe Zone’ Plan to T...   
23479  How to Blow $700 Million: Al Jazeera America F...   
23480  10 U.S. Navy Sailors Held by Iranian Military ...   

                                                    text       subject  \
0      WASHINGTON (Reuters) - The head of a conservat...  politicsNews   
1      WASHINGTON (Reuters) - Transgender people will...  politicsNews   
2      WASHINGTON (Reuters) - The special counsel inv...  politicsNews   
3      WASHINGTON (Reuters) - Trump campaign adviser ...  politicsNews   
4      SEATTLE/WASHINGTON (Reuters) - President Donal...  politicsNews   
...                                                  ...           ...   
23476  21st Century Wire says As 21WIRE reported earl...   Middle-east   
23477  21st Century Wire says It s a familiar theme. ...   Middle-east   
23478  Patrick Henningsen  21st Century WireRemember ...   Middle-east   
23479  21st Century Wire says Al Jazeera America will...   Middle-east   
23480  21st Century Wire says As 21WIRE predicted in ...   Middle-east   

                     date statement  
0      December 31, 2017       True  
1      December 29, 2017       True  
2      December 31, 2017       True  
3      December 30, 2017       True  
4      December 29, 2017       True  
...                   ...       ...  
23476    January 16, 2016      fake  
23477    January 16, 2016      fake  
23478    January 15, 2016      fake  
23479    January 14, 2016      fake  
23480    January 12, 2016      fake  

[44898 rows x 5 columns]

In [172]:
data.reset_index(inplace=True)

In [173]:
data.drop(columns='index',inplace=True)

In [174]:
data.rename(columns={'R':'Statement'},inplace=True)

In [175]:
data

title  \
0      As U.S. budget fight looms, Republicans flip t...   
1      U.S. military to accept transgender recruits o...   
2      Senior U.S. Republican senator: 'Let Mr. Muell...   
3      FBI Russia probe helped by Australian diplomat...   
4      Trump wants Postal Service to charge 'much mor...   
...                                                  ...   
44893  McPain: John McCain Furious That Iran Treated ...   
44894  JUSTICE? Yahoo Settles E-mail Privacy Class-ac...   
44895  Sunnistan: US and Allied ‘Safe Zone’ Plan to T...   
44896  How to Blow $700 Million: Al Jazeera America F...   
44897  10 U.S. Navy Sailors Held by Iranian Military ...   

                                                    text       subject  \
0      WASHINGTON (Reuters) - The head of a conservat...  politicsNews   
1      WASHINGTON (Reuters) - Transgender people will...  politicsNews   
2      WASHINGTON (Reuters) - The special counsel inv...  politicsNews   
3      WASHINGTON (Reuters) - Trump campaign adviser ...  politicsNews   
4      SEATTLE/WASHINGTON (Reuters) - President Donal...  politicsNews   
...                                                  ...           ...   
44893  21st Century Wire says As 21WIRE reported earl...   Middle-east   
44894  21st Century Wire says It s a familiar theme. ...   Middle-east   
44895  Patrick Henningsen  21st Century WireRemember ...   Middle-east   
44896  21st Century Wire says Al Jazeera America will...   Middle-east   
44897  21st Century Wire says As 21WIRE predicted in ...   Middle-east   

                     date statement  
0      December 31, 2017       True  
1      December 29, 2017       True  
2      December 31, 2017       True  
3      December 30, 2017       True  
4      December 29, 2017       True  
...                   ...       ...  
44893    January 16, 2016      fake  
44894    January 16, 2016      fake  
44895    January 15, 2016      fake  
44896    January 14, 2016      fake  
44897    January 12, 2016      fake  

[44898 rows x 5 columns]

###### The final dataset before Text preprocessing

In [176]:
import re
import emoji
from textblob import TextBlob
from nltk.tokenize import sent_tokenize,word_tokenize
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer,LancasterStemmer,SnowballStemmer,WordNetLemmatizer

###### 1.Here we are using `regex` for removing unnecessary characters in text 
###### 2. `emoji` using this we are decoding the information from emoji present in the text
###### 3. `TextBlob` using textblob we are correcting the spelling of english words
###### 4. `Stem` using stemming we are storing the root word of a particular word 
###### 5. `tokenize` using this dividing paragraph or sentences  into tokens

In [177]:
stp=stopwords.words('english')

In [178]:
'not' in stp

True

In [179]:
stp.remove('not')

###### removing `not` in the stopwords 

In [180]:
new0=[]
for sent in data['text']:
    new1=[]
    new=sent.lower()              #first converting into lower case
    new=emoji.demojize(new)       #getting desired information from the emojis
    new=TextBlob(new).correct().string    #correcting the spelling mistakes 
    new=re.sub('<.*?>','',new)            # removing unnecessary characters
    new=re.sub('https?://\S*|www\S*','',new)
    new=re.sub("[^a-z ]","",new)
    for word in word_tokenize(new):      #using word_tokenize dividing into words
        if word in stp:                  #neglecting the words which are in the stopwords 
            pass
        else:
            stem=WordNetLemmatizer().lemmatize(word) #stemming of each word which are unique
            new1.append(stem)
    new0.append(' '.join(new1))                     #Appending into the list

In [181]:
cdata=pd.DataFrame({'cleaned_text':new0})

In [182]:
cdata

cleaned_text
0      washington reuters head conservative republica...
1      washington reuters transgender people allowed ...
2      washington reuters special counsel investigati...
3      washington reuters trump campaign adviser geor...
4      seattlewashington reuters president donald tru...
...                                                  ...
44893  st century wire say wire reported earlier week...
44894  st century wire say familiar theme whenever di...
44895  patrick henningsen st century wireremember oba...
44896  st century wire say al jazeera america go hist...
44897  st century wire say wire predicted new year lo...

[44898 rows x 1 columns]

###### The final dataset After Text preprocessing

# BOW

In [111]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

###### using `CountVectorizer` converting text into vectors

In [112]:
bow=CountVectorizer()

In [113]:
spar=bow.fit_transform(cdata['cleaned_text'])

In [119]:
spar

<44898x204269 sparse matrix of type '<class 'numpy.int64'>'
	with 7072108 stored elements in Compressed Sparse Row format>

In [128]:
spar[0:5]

<5x204269 sparse matrix of type '<class 'numpy.int64'>'
	with 1128 stored elements in Compressed Sparse Row format>

###### Due to getting `MemoryError` we are considering only `5` rows of sparse matrix

In [115]:
bow.vocabulary_

{'washington': 195497,
 'reuters': 152623,
 'head': 74053,
 'conservative': 35821,
 'republican': 151249,
 'faction': 58198,
 'congress': 35501,
 'voted': 194076,
 'month': 112726,
 'huge': 81944,
 'expansion': 57017,
 'national': 116076,
 'debt': 42769,
 'pay': 129300,
 'tax': 176934,
 'cut': 40573,
 'called': 25009,
 'fiscal': 61423,
 'sunday': 174065,
 'urged': 190814,
 'budget': 23388,
 'restraint': 152211,
 'keeping': 94026,
 'sharp': 163238,
 'pivot': 136808,
 'way': 196061,
 'among': 7420,
 'representative': 151150,
 'mark': 105824,
 'meadow': 107781,
 'speaking': 168869,
 'cbs': 27625,
 'face': 57890,
 'nation': 116073,
 'drew': 49801,
 'hard': 73141,
 'line': 101013,
 'federal': 59850,
 'spending': 169180,
 'lawmaker': 98320,
 'bracing': 21815,
 'battle': 15755,
 'january': 90472,
 'return': 152519,
 'holiday': 77676,
 'wednesday': 196633,
 'begin': 16542,
 'trying': 186329,
 'pas': 128750,
 'fight': 60634,
 'likely': 100811,
 'linked': 101120,
 'issue': 89053,
 'immigration':

In [116]:
import numpy as np

In [130]:
cd=pd.DataFrame(spar[0:5].toarray(),columns=sorted(bow.vocabulary_.keys()))

In [134]:
cd

aa  aaa  aaaaackkk  aaaahhhh  aaaarrgh  \
0   0    0          0         0         0   
1   0    0          0         0         0   
2   0    0          0         0         0   
3   0    0          0         0         0   
4   0    0          0         0         0   

   aaainnertextreplacernrngquotnquotelsevar  aaarf  aab  aaba  aabo  ...  \
0                                         0      0    0     0     0  ...   
1                                         0      0    0     0     0  ...   
2                                         0      0    0     0     0  ...   
3                                         0      0    0     0     0  ...   
4                                         0      0    0     0     0  ...   

   zypries  zyries  zyuganov  zyuseragentifzxzbreak  zz  zzbluecomet  zztaine  \
0        0       0         0                      0   0            0        0   
1        0       0         0                      0   0            0        0   
2        0       0         0                      0   0            0        0   
3        0       0         0                      0   0            0        0   
4        0       0         0                      0   0            0        0   

   zzzzaaaacccchhh  zzzzzzzz  zzzzzzzzzzzzz  
0                0         0              0  
1                0         0              0  
2                0         0              0  
3                0         0              0  
4                0         0              0  

[5 rows x 204269 columns]

###### Converted into vectors of having `204269`  Dimensions

In [131]:
cd.columns

Index(['aa', 'aaa', 'aaaaackkk', 'aaaahhhh', 'aaaarrgh',
       'aaainnertextreplacernrngquotnquotelsevar', 'aaarf', 'aab', 'aaba',
       'aabo',
       ...
       'zypries', 'zyries', 'zyuganov', 'zyuseragentifzxzbreak', 'zz',
       'zzbluecomet', 'zztaine', 'zzzzaaaacccchhh', 'zzzzzzzz',
       'zzzzzzzzzzzzz'],
      dtype='object', length=204269)

# Tf-Idf

In [132]:
tfidf=TfidfVectorizer()

In [133]:
spar=tfidf.fit_transform(cdata['cleaned_text'])

In [135]:
tfidf.vocabulary_

{'washington': 195497,
 'reuters': 152623,
 'head': 74053,
 'conservative': 35821,
 'republican': 151249,
 'faction': 58198,
 'congress': 35501,
 'voted': 194076,
 'month': 112726,
 'huge': 81944,
 'expansion': 57017,
 'national': 116076,
 'debt': 42769,
 'pay': 129300,
 'tax': 176934,
 'cut': 40573,
 'called': 25009,
 'fiscal': 61423,
 'sunday': 174065,
 'urged': 190814,
 'budget': 23388,
 'restraint': 152211,
 'keeping': 94026,
 'sharp': 163238,
 'pivot': 136808,
 'way': 196061,
 'among': 7420,
 'representative': 151150,
 'mark': 105824,
 'meadow': 107781,
 'speaking': 168869,
 'cbs': 27625,
 'face': 57890,
 'nation': 116073,
 'drew': 49801,
 'hard': 73141,
 'line': 101013,
 'federal': 59850,
 'spending': 169180,
 'lawmaker': 98320,
 'bracing': 21815,
 'battle': 15755,
 'january': 90472,
 'return': 152519,
 'holiday': 77676,
 'wednesday': 196633,
 'begin': 16542,
 'trying': 186329,
 'pas': 128750,
 'fight': 60634,
 'likely': 100811,
 'linked': 101120,
 'issue': 89053,
 'immigration':

In [140]:
cd1=pd.DataFrame(spar[0:5].toarray(),columns=sorted(tfidf.vocabulary_.keys())).applymap(int)

In [141]:
cd1

aa  aaa  aaaaackkk  aaaahhhh  aaaarrgh  \
0   0    0          0         0         0   
1   0    0          0         0         0   
2   0    0          0         0         0   
3   0    0          0         0         0   
4   0    0          0         0         0   

   aaainnertextreplacernrngquotnquotelsevar  aaarf  aab  aaba  aabo  ...  \
0                                         0      0    0     0     0  ...   
1                                         0      0    0     0     0  ...   
2                                         0      0    0     0     0  ...   
3                                         0      0    0     0     0  ...   
4                                         0      0    0     0     0  ...   

   zypries  zyries  zyuganov  zyuseragentifzxzbreak  zz  zzbluecomet  zztaine  \
0        0       0         0                      0   0            0        0   
1        0       0         0                      0   0            0        0   
2        0       0         0                      0   0            0        0   
3        0       0         0                      0   0            0        0   
4        0       0         0                      0   0            0        0   

   zzzzaaaacccchhh  zzzzzzzz  zzzzzzzzzzzzz  
0                0         0              0  
1                0         0              0  
2                0         0              0  
3                0         0              0  
4                0         0              0  

[5 rows x 204269 columns]

In [ ]:
:)